### PR TITLE
Introduce C++ wrappers for gpr_mu and gpr_cv.

### PR DESCRIPTION
**Pull request by soheilhy**
Introduce RAII wrappers in the grpc::internal and grpc_core
namespaces, and use them in place of std::mutex and
std::condition_variable.

Note that, since std::mutex is also used by the public
C++ headers we cannot introduce these wrappers in grpc_core.

Also, note that in grpcpp we cannot use gRPC core and vice versa.
So we had to duplicate the code, once using core_codegen_interface
and once using direct calls.

This reapplies https://github.com/grpc/grpc/pull/17932
with a union in include/grpcpp/impl/codegen/sync.h to make sure
we have at least 40B reserved.

### CODE REVIEW

**Comment by soheilhy** [comment]
This is just to try the upstream tests.

**Comment by markdroth** [comment]
Can you recreate this as one commit that is a revert of the previous revert and then a second commit showing the subsequent changes?  That will make it much easier to review.

**Comment by soheilhy** [comment]
Sure, please see the diff in `https://github.com/grpc/grpc/pull/18774/commits/b87454ccb4c2f58b10d9bf6e801491a6e93cc50d`

File: include/grpcpp/impl/codegen/sync.h [hunk 1]
```diff
@@ -19,8 +19,15 @@
 #ifndef GRPCPP_IMPL_CODEGEN_SYNC_H
 #define GRPCPP_IMPL_CODEGEN_SYNC_H
 
-#include <grpc/impl/codegen/log.h>
 #include <grpc/impl/codegen/port_platform.h>
+
+#ifndef GPR_WINDOWS
```
Comment (markdroth) [inline include/grpcpp/impl/codegen/sync.h]: Hmm... Do we want this kind of conditional outside of port_platform.h?

@nicolasnoble should weigh in on this.

Comment (soheilhy) [reply include/grpcpp/impl/codegen/sync.h]: Here, we have to include `<pthread>` directly without relying on `GPR_SYNC_POSIX` to make sure  the space is large enough wherever `pthread_mutex_t` can possibly be use in place of `gpr_mu`.

**Comment by soheilhy** [comment]
Thank you for the approval @markdroth !

**Comment by soheilhy** [comment]
Thank you @nicolasnoble for the suggestion!

**Comment by soheilhy** [comment]
Windows failure are the flakes in #18637
I re-triggered the builds.

**Pull request merged**